### DISCUSSION TITLE
General purpose PCM channel idea

### DISCUSSION THREAD

**Discussion by Weeppiko** [nesbox/TIC-80#2487] [Ideas]
Currently the only way to play PCM data on TIC-80 it's through the `Sound Registers`, but the sound suffers a lot from this, since it is only able to play up to 1920hz and there is only 4 bits of depth, and on top of that there is also the fact that some flat lines are threated as noise, which makes it even worse to deal with. There was a mention about being able to poke to the sound registers during `SCN` to help with that but this was never really addressed.

So, I was thinking about how else it could be done, and the idea was to have only a small buffer in `RAM` (no pitch control or whatever) that would be fetched every frame and played back, e.g. having 128 bytes of samples would result in a sound quality of 7680hz, which is *way* better than what is currently possible, but still sounds retro enough, and can be used for any audio related stuff, like custom synths, ability to play voice clips for sfx, generative audio stuff, etc.

**Comment by nesbox** [comment]
I think it's a good idea, I implemented PCM support on this branch https://github.com/nesbox/TIC-80/tree/pcm, the binaries will be soon here https://github.com/nesbox/TIC-80/actions/runs/8402082845

Could someone test it and give us feedback?
To write to the PCM buffer please use `poke(0x14E24, value)` command.
![image](https://github.com/nesbox/TIC-80/assets/1101448/2d42f1fe-686f-46e5-ba98-6ecf62c11b69)

**Reply by Weeppiko** [reply]
I did a small test with a saw wave and noticed it doesn't reach the bottom part of the phase, I suppose the value can be threated as a signed byte instead (-128 to 127)
![image](https://github.com/nesbox/TIC-80/assets/88540876/5fb93434-7487-41e5-98c6-f66a6f68a0fa)

**Reply by nesbox** [reply]
Please try to check the build from the main branch, because I already merged the PCM feature where I used 0-255 values.
Thanks.

**Reply by Weeppiko** [reply]
Sounded a bit quieter, but I think it is fine overall.

**Comment by nesbox** [comment]
A very simple example on JS to generate a sin wave using PCM
```js
// script:js

let t=0
function TIC()
{
	cls(13)
	for(let i=0;i<128;i++)
		poke(0x14E24+i,(Math.sin(t*128+i)+1)*128)

	t++
}
```

**Comment by MineRobber9000** [comment]
So what sample rate are we talking about here? 128 samples per frame * 60 frames per second would indicate 7680Hz?

**EDIT:** nevermind, just read the original post. nice!

**Comment by Weeppiko** [comment]
In case anyone wondering how it sounds: 
[new_pcm.zip](https://github.com/nesbox/TIC-80/files/14733788/new_pcm.zip)

**Comment by HanamileH** [comment]
This has just invalidated all my work on creating an aPCM that tries to produce the most perfect sound possible using sound registers ,_,

**Reply by Skeptim** [reply]
That doesn't make it any less of a feat ^^

**Reply by nesbox** [reply]
I don’t think this invalidated your awesome work, I think people won’t be able to load a lot of data into the PCM channel to play music, but it definitely increases the scope for creativity using generative audio and etc.

**Comment by speedoog** [comment]
@nesbox I've heard the result and it's really great :)
Do you know when it would be available on the official version ?

**Comment by Luke-Software1024** [comment]
Here's a demo of this functionality:
https://tic80.com/play?cart=4744
